### ISSUE TITLE
[RyuJIT] bounds checks aren't eliminated for array[0] access when different indices are already accessed

### ISSUE THREAD

**Issue by EgorBo** [dotnet/runtime#39156]
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static byte GetValue() => 42;

static void Test1(Span<byte> span)
{
    span[2] = GetValue();  // bounds check
    span[1] = GetValue(); 
}

static void Test2(Span<byte> span)
{
    span[2] = GetValue();  // bounds check
    span[0] = GetValue();  // bounds check (redundant!)
}
```
#### Codegen:
```asm
; Method Program:Test1(System.Span`1[Byte])
G_M8936_IG01:
       57                   push     rdi
       56                   push     rsi
       4883EC28             sub      rsp, 40
G_M8936_IG02:
       488B31               mov      rsi, bword ptr [rcx]
       8B4108               mov      eax, dword ptr [rcx+8]
G_M8936_IG03:
       83F802               cmp      eax, 2
       761C                 jbe      SHORT G_M8936_IG05
       488D7E02             lea      rdi, bword ptr [rsi+2]
       E86EA6FEFF           call     Program:GetValue():ubyte
       8807                 mov      byte  ptr [rdi], al
       48FFC6               inc      rsi
       E864A6FEFF           call     Program:GetValue():ubyte
       8806                 mov      byte  ptr [rsi], al
G_M8936_IG04:
       4883C428             add      rsp, 40
       5E                   pop      rsi
       5F                   pop      rdi
       C3                   ret      
G_M8936_IG05:
       E82EAB915F           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3     
; Total bytes of code: 51


; Method Program:Test2(System.Span`1[Byte])
G_M23691_IG01:
       57                   push     rdi
       56                   push     rsi
       53                   push     rbx
       4883EC20             sub      rsp, 32
G_M23691_IG02:
       488B31               mov      rsi, bword ptr [rcx]
       8B7908               mov      edi, dword ptr [rcx+8]
G_M23691_IG03:
       83FF02               cmp      edi, 2
       761F                 jbe      SHORT G_M23691_IG05
       488D5E02             lea      rbx, bword ptr [rsi+2]
       E81DA6FEFF           call     Program:GetValue():ubyte
       8803                 mov      byte  ptr [rbx], al
       83FF00               cmp      edi, 0                   ;;;  <---- redundant bounds check
       760F                 jbe      SHORT G_M23691_IG05      ;;;  <---- redundant bounds check
       E811A6FEFF           call     Program:GetValue():ubyte
       8806                 mov      byte  ptr [rsi], al
G_M23691_IG04:
       4883C420             add      rsp, 32
       5B                   pop      rbx
       5E                   pop      rsi
       5F                   pop      rdi
       C3                   ret      
G_M23691_IG05:
       E8DAAA905F           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3     
; Total bytes of code: 55
```

My guess it's "dead store removal" kills GT_ASG with GT_COMMA for `span[0]` and leaves only lonely ARR_BOUNDS_CHECK which then is not handled by optRemoveRangeCheck.
```

BB02 - Dead assignment has side effects...
N007 ( 11, 13) [000034] -A-XG---R---              *  ASG       byref 
N006 (  3,  2) [000033] D------N----              +--*  LCL_VAR   byref  V03 tmp2         
N005 (  7, 10) [000031] ---XG-------              \--*  COMMA     byref 
N003 (  6,  9) [000026] ---XG-------                 +--*  ARR_BOUNDS_CHECK_Rng void  
N001 (  1,  1) [000021] ------------                 |  +--*  CNS_INT   int    0
N002 (  1,  1) [000025] ------------                 |  \--*  LCL_VAR   int    V05 tmp4         
N004 (  1,  1) [000029] ------------                 \--*  LCL_VAR   byref  V04 tmp3         

                                                        ↓   ↓   ↓

N003 (  6,  9) [000026] ---XG-------              *  ARR_BOUNDS_CHECK_Rng void  
N001 (  1,  1) [000021] ------------              +--*  CNS_INT   int    0
N002 (  1,  1) [000025] ------------              \--*  LCL_VAR   int    V05 tmp4    
```

Possible fix (works): https://github.com/EgorBo/runtime-1/commit/a8367b83c08fcc6ac4f48c5e1a16881c49927dc5
Jit-diff:
```
Total bytes of diff: -563 (-0.001% of base)
    diff is an improvement.

Top file improvements (bytes):
        -134 : System.Numerics.Tensors.dasm (-0.034% of base)
         -96 : System.Net.Security.dasm (-0.068% of base)
         -76 : System.Net.Http.dasm (-0.009% of base)
         -74 : System.Private.CoreLib.dasm (-0.001% of base)
         -68 : System.Net.Mail.dasm (-0.032% of base)
         -56 : System.Net.HttpListener.dasm (-0.024% of base)
         -28 : System.Text.RegularExpressions.dasm (-0.011% of base)
         -24 : Microsoft.CSharp.dasm (-0.006% of base)
          -7 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (-0.000% of base)

9 total files with Code Size differences (9 improved, 0 regressed), 255 unchanged.
```

/cc @dotnet/jit-contrib 

category:cq
theme:bounds-checks
skill-level:intermediate
cost:medium

**Comment by erozenfeld** [comment]
@EgorBo Perhaps a better fix would be to teach `gtExtractSideEffList` to keep the parent `COMMA` for `ARR_BOUNDS_CHECK`.

**Comment by SingleAccretion** [comment]
Fixed in #49271.

**Comment by SingleAccretion** [comment]
Closing as per the above comment.
